### DISCUSSION TITLE
Expose collected metrics on `/q/metrics` endpoint

### DISCUSSION THREAD

**Discussion by ShubhamRwt** [quarkusio/quarkus#32647] [Q&A] [answered]
Hi,
Based on the chats on the previous [discussion](https://github.com/quarkusio/quarkus/discussions/32639#discussion), we decided move forward using the JMX  to scrape the kafka related metrics.
I am currently using Prometheus JmXCollector class to collect metrics from JMX. I wanted to know how we can expose these collected metrics on the same /q/metrics endpoint where micrometer metrics are exposed?

CC: @ppatierno

[a comment by an automated account is not shown]

**Comment by ebullient** [comment]
There is a separate Quarkus JMX Micrometer registry present in the Quarkiverse project:

https://github.com/quarkiverse/quarkus-micrometer-registry

https://quarkiverse.github.io/quarkiverse-docs/quarkus-micrometer-registry/dev/micrometer-registry-jmx.html

You can have both registries enabled at the same time, but no, the JMX registry does not use the metrics URL. JMX is a different animal.

**Reply by ShubhamRwt** [reply]
So it is not possible to inject our collected metrics onto `/q/metrics` endpoint?

**Reply by brunobat** [reply]
You want to have particular JMX produced metrics on the micrometer `/q/metrics` endpoint?

**Reply by ShubhamRwt** [reply]
Exactly @brunobat

**Reply by ppatierno** [reply]
@ebullient @brunobat we don't want to expose a JMX endpoint on which making available some metrics.
We are collecting metrics from Kafka clients (they expose the metrics via JMX) and then we want to expose those metrics on the same `/q/metrics` endpoint in Prometheus format.
To provide you more information about how it's working today see this:

https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/java/io/strimzi/kafka/bridge/JmxCollectorRegistry.java

The `scrape` is called everytime the current `/metrics` endpoint is queried. So it collects JMX metrics via the JMX collector and return the Prometheus string format.
This is happening inside this overall reporter here:

https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/java/io/strimzi/kafka/bridge/MetricsReporter.java#L62

where other than scraping JMX we are also returning the metrics from the default Micrometer registry provided via Vert.x (with all its HTTP server related metrics out of box).

Finally this is where we have an handler for the `/metrics` route which is called to scrape as above code:

https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java#L182

We want to reproduce a kind of same result: Quarkus is providing us the HTTP server metrics, we want add our Kafka metrics and making them available on the same `/q/metrics` endpoint.

**Reply by ebullient** [reply]
Hmm. the way that is implemented, no. It is two separate data aggregations... The micrometer prometheus registry (this is not a quarkus thing) is collecting metrics from a higher level abstraction.. and you've already gone to a specific prometheus collector in strimzi.. 

well... no. hang on. 

you could create a new (not-default) prometheus MeterRegistry that injects the one from strimzi... it would have to do some kind of munging between the two prometheus collectors at scrape time

**Reply by ppatierno** [reply]
> you could create a new (not-default) prometheus MeterRegistry that injects the one from strimzi... it would have to do some kind of munging between the two prometheus collectors at scrape time

@ebullient can you elaborate a little bit more this idea? I have played around with what we have now and how to deal with the `PrometheusMeterRegistry` instance (injected via Quarkus extension) together with the `JmxCollector` but didn't find a way to converge metrics from the two different sources in one place.

The hack that I started to have in mind was about using an HTTP filter, to intercept the call on the `/q/metrics` endpoint and doing the same scraping we do today this way:

```java
public String scrape() {
        StringBuilder sb = new StringBuilder();
        if (jmxCollectorRegistry != null) {
            sb.append(jmxCollectorRegistry.scrape());
        }
        if (meterRegistry instanceof PrometheusMeterRegistry) {
            sb.append(((PrometheusMeterRegistry) meterRegistry).scrape());
        }
        return sb.toString();
    }
```

but it seems that Quarkus doesn't hit any `@ServerRequestFilter` when it's a non-application path (i.e. '/q/metrics`) :-(

**Reply by ppatierno** [reply]
@ebullient I decided to go the same way we go today to get metrics from the two worlds (the MeterRegistry and JMX collector).
I defined our own `/metrics` endpoint (which is the one we have in the current bridge version) this way:

https://github.com/ppatierno/strimzi-kafka-bridge/blob/add-metrics-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java#L353

then that `MetricsReporter` component is just "scraping" on both JMX collector and the Quarkus injected `MeterRegistry` and then builds the output text:

https://github.com/ppatierno/strimzi-kafka-bridge/blob/add-metrics-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/MetricsReporter.java#L38

The only drawnback is that the `/q/metrics` endpoint is still active but getting just the Quarkus metrics. I would like to "override" that endpoint (so our code to be called), or "suppress" or just mapping it to our `/metrics` endpoint and our code still executed.
I haven't found a good way to do that, so I opened this discussion https://github.com/quarkusio/quarkus/discussions/32671

**Reply by ebullient** [reply, marked as answer]
So. What I was trying to suggest is that you replace the default Prometheus Meter Registry with your own..

The default producer will be replaced if you make your own: 

https://github.com/quarkusio/quarkus/blob/b687fbb7ba8678d99433be71aca5dd1f358fb6e6/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/PrometheusMeterRegistryProducer.java#L20

You can also disable construction of the default Prometheus Meter Registry with a property (assuming you add your own instead): 

`quarkus.micrometer.export.prometheus.default-registry=false`

As shown here (with StatsD):

https://quarkus.io/guides/micrometer#creating-a-customized-meterregistry

My thought was that you could extend the PrometheusMeterRegistry (which would catch micrometer metrics) and add what you need to merge in the other metrics. Given you have the producer, you can inject sources and control construction, and then you'd need to override the scrape method that accepts the content type: 

https://github.com/micrometer-metrics/micrometer/blob/23804d52a78cc1ce4dd248169c76cf59f0f9d7ad/implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheus/PrometheusMeterRegistry.java#L117

The default handler just calls that scrape method and gets a string back (it doesn't otherwise know what happens): 

https://github.com/quarkusio/quarkus/blob/b687fbb7ba8678d99433be71aca5dd1f358fb6e6/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/handlers/PrometheusHandler.java#L63

**Reply by ShubhamRwt** [reply]
Hi @ebullient I was trying to implement what you explained. When you say `extend the PrometheusMeterRegistry` you want me to do something like this --> `public CustomMeterRegistryProducer extends PrometheusMeterRegistry`, right and also then overriding the scrape as per our requirement? (CustomMeterRegistryProducer is the class I am creating)

**Comment by ShubhamRwt** [comment]
Hi @ebullient. I am now able to get the required metrics now on `/q/metrics` endpoint but yeah they are always enable. Anyway I can disable them?

**Reply by ShubhamRwt** [reply]
Wait, I guess the jmx ones are not showing up. Let me get back here after confirming

**Reply by ShubhamRwt** [reply]
I think my overriden `scrape()` method is not being triggered

**Reply by ebullient** [reply]
so now we just need to verify that your registry bean is being used, there are several scrape methods.. so make sure you're overriding the correct one

**Reply by ShubhamRwt** [reply]
https://gist.github.com/ShubhamRwt/fc7ff38ef1599467f7168ba623919d2a Can you have a look and see if I am going wrong somewhere. Foe ease I am just trying to create `PrometheusMeteRegistry` and added the code for jmx scraping in `scrape` method

**Reply by ShubhamRwt** [reply]
@ebullient ^

**Reply by ShubhamRwt** [reply]
@ebullient We are using quarkus 2.15.3 and there the `doHandler` method is like this
```java
    private void doHandle(HttpServerResponse response) {
        response.putHeader("Content-Type", TextFormat.CONTENT_TYPE_004)
                .end(Buffer.buffer(registry.scrape()));
    }
```

Also I have `quarkus.micrometer.export.prometheus.default-registry=false` added this to my properties file and still the metrics are there on `/q/metrics`

**Reply by ShubhamRwt** [reply]
Also I found out if I do this 
```java     
    @Produces
    @Singleton
    @Alternative
    public PrometheusMeterRegistry createPrometheusMeterRegistry() {
        return new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
    }
```
I get the `jmx` scraped metrics but not the http ones

But if I do this 
```java
    @Produces
    @Singleton
    @Alternative
    @Priority(Interceptor.Priority.APPLICATION + 100)
    public PrometheusMeterRegistry createPrometheusMeterRegistry() {
        return new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
    }
``` 
I get the `http` ones but not the jmx ones

**Comment by ShubhamRwt** [comment]
@ebullient Thanks for the instructions. Iam now able to get both the metrics on `/q/metrics` endpoint.

**Reply by brunobat** [reply]
Cool, this comment did it then: https://gist.github.com/ShubhamRwt/fc7ff38ef1599467f7168ba623919d2a?permalink_comment_id=4541388#gistcomment-4541388
(For the future)

**Reply by ShubhamRwt** [reply]
Yeah, but instead of   `sb.append(this.scrape());`,   I am using   sb.append(super.scrape()); since `this.scrape()` will call the overriden `scrape()` method which we don't want for the http metrics

**Reply by ShubhamRwt** [reply]
@brunobat I have marked the procedure told above by @ebullient as the answer

**Reply by ShubhamRwt** [reply]
@brunobat @ebullient Is it somehow possible that if we call `/metrics` then it is forwarded to `/q/metrics`?

**Reply by ebullient** [reply]
you can change the metrics endpoint from /q/metrics to /metrics if you want. That's a normal configuration property

**Reply by ShubhamRwt** [reply]
Thanks @ebullient for clarifying all my doubts. Is there any documentation/example which I can refer to for changing the endpoint?

**Reply by ShubhamRwt** [reply]
@ebullient got the property. Thanks for all the help.

**Reply by ShubhamRwt** [reply]
Sorry forgot that there was grep being applied

**Reply by ShubhamRwt** [reply]
Thanks a lot again @ebullient

**Comment by ShubhamRwt** [comment]
I will be closing this discussion
